server/s3_copy_object_action.cc

```diff
@@ -76,10 +77,26 @@ void S3CopyObjectAction::setup_steps() {
 void S3CopyObjectAction::get_source_bucket_and_object() {
   s3_log(S3_LOG_DEBUG, request_id, "%s Entry\n", __func__);
   std::string source = request->get_headers_copysource();
-  size_t separator_pos = source.find("/");
-  if (separator_pos != std::string::npos) {
-    source_bucket_name = source.substr(0, separator_pos);
-    source_object_name = source.substr(separator_pos + 1);
+  size_t separator_pos;
+  if (source[0] != '/') {
+    separator_pos = source.find("/");
+    if (separator_pos != std::string::npos) {
+      source_bucket_name = source.substr(0, separator_pos);
+      source_object_name = source.substr(separator_pos + 1);
+      s3_log(S3_LOG_DEBUG, request_id,
+             "If block, bucket name = %s, object name = %s\n",
+             source_bucket_name.c_str(), source_object_name.c_str());
+    }
+  } else {
+    separator_pos = source.find("/", 1);
+    if (separator_pos != std::string::npos) {
+      source_bucket_name = source.substr(1, separator_pos - 1);
+      source_object_name = source.substr(separator_pos + 1);
+      source_object_name = url_decode(source_object_name.c_str());
```

Comment:
> why url_decode only in else block, not in if block ?

Because the if block is where we seem to be getting the request from CLI, and it does not contain the char encoding, but the else block is coming from UI and here when there is an object inside a folder, the slash is encoded to '%2F'. So url_decode is a must.


> Also how the problem of unicode chars is solved for other APIs, like GET or PUT ?

Not sure, couldn't find any reference to url_decode but url_encode already exists.